In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score, mean_squared_error
from sklearn.preprocessing import MultiLabelBinarizer
import seaborn as sns
import os
import re
import spacy
from datetime import datetime, date
nlp = spacy.load('en')
from collections import Counter
%matplotlib inline

In [3]:
list_df = pd.read_csv('./archive/listings.csv')
cal_df = pd.read_csv('./archive/calendar.csv')
review_df = pd.read_csv('./archive/reviews.csv')

We want to determine what features a listing must have to generate the highest price point. Features examined will be neighbourhood, amenities included, and rating

- What style of accommodation  allows for highest revenue?
- Can we predict price using categorical and quantitative values?
- What effect does distance from city center have on price?

The listings dataset contains the fields of interest in analyzing the dataset. The calendar and reviews csv files are not needed for the purpose of this project. The following cell analyzed the size and data types contained in the file.

In [4]:
print('Shape:', list_df.shape)
print('Data types column count:')
print(list_df.dtypes.value_counts())

Shape: (3818, 92)
Data types column count:
object     62
float64    17
int64      13
dtype: int64


The dataset contains entries for 3818 listings with 92 variables per listing. These variables are represented by a mix of object, float64, and int64 data types. This indicates that a mix of quantitative and categorical variables can be found for each listing, complicating our analysis of the data.

Now that the types of data contained in the dataset are known, it is important to analyze missing values in the dataset. This will aid in determining which columns might not be useful in our analysis due to a lack of entries.

In [554]:
#Count the number of missing values and then sort in descending order
pd.set_option("max_rows", None)
list_df.isna().sum().sort_values(ascending=False)

license                             3818
square_feet                         3721
monthly_price                       2301
security_deposit                    1952
weekly_price                        1809
notes                               1606
neighborhood_overview               1032
cleaning_fee                        1030
transit                              934
host_about                           859
host_acceptance_rate                 773
review_scores_accuracy               658
review_scores_checkin                658
review_scores_value                  656
review_scores_location               655
review_scores_cleanliness            653
review_scores_communication          651
review_scores_rating                 647
reviews_per_month                    627
first_review                         627
last_review                          627
space                                569
host_response_time                   523
host_response_rate                   523
neighbourhood   

The missing values in some of these columns will need to be dealt with to accurately analyze the data. Certain columns containing missing values are not of interest, contain over 50 % missing values without an explanation, or are categorical variables containing a written description that is not within the scope of our analysis. These columns dropped completely (license, notes, neighborhood_overview, transit, host_about, first_review, last_review, space, any of the url vairables, summary, host_location). Monthly and weekly price variables will be dropped considering the price variable encompasses these values and is more universally listed in the dataset. Security_deposit, cleaning_fee, and extra_people are all columns that contain fee information where an absent value is likely an indication of a zero $ fee and the missing value will be replaced with 0. All other columns with missing values will have the missing entries filled with mean or mode values.

In [556]:
list_df.security_deposit

0             NaN
1         $100.00
2       $1,000.00
3             NaN
4         $700.00
5             NaN
6         $150.00
7         $150.00
8             NaN
9         $100.00
10            NaN
11        $500.00
12        $300.00
13        $400.00
14      $1,500.00
15            NaN
16        $100.00
17        $100.00
18            NaN
19        $500.00
20            NaN
21            NaN
22            NaN
23            NaN
24        $500.00
25        $500.00
26        $250.00
27        $150.00
28        $500.00
29            NaN
30            NaN
31        $500.00
32        $200.00
33        $500.00
34        $250.00
35            NaN
36        $750.00
37            NaN
38            NaN
39            NaN
40        $100.00
41            NaN
42        $500.00
43            NaN
44            NaN
45            NaN
46            NaN
47            NaN
48        $100.00
49        $140.00
50            NaN
51            NaN
52            NaN
53            NaN
54        $500.00
55        

### Quantitative Transformations

List of variables

In [518]:
quant_cols = ['host_since', 'host_response_rate', 'host_acceptance_rate',
       'host_listings_count', 'host_total_listings_count', 'accommodates',
       'bathrooms', 'bedrooms', 'beds', 'square_feet', 'price', 'security_deposit',
       'cleaning_fee', 'guests_included', 'extra_people', 'minimum_nights',
       'maximum_nights', 'availability_30', 'availability_60', 'availability_90',
       'availability_365', 'number_of_reviews', 'first_review', 'last_review', 'review_scores_rating',
       'review_scores_accuracy', 'review_scores_cleanliness',
       'review_scores_checkin', 'review_scores_communication',
       'review_scores_location', 'review_scores_value', 'reviews_per_month']

cat_cols = ['host_response_time','host_is_superhost', 'host_verifications', 'host_has_profile_pic', 'host_identity_verified',
       'property_type', 'room_type', 'bed_type', 'amenities', 'instant_bookable',
       'cancellation_policy', 'require_guest_profile_picture', 'require_guest_phone_verification', 
       'neighbourhood_cleansed', 'city', 'state', 'country_code', 'requires_license', 'is_location_exact',
       'has_availability']

In [520]:
date_cols = ['host_since', 'first_review', 'last_review']
value_cols = ['host_response_rate', 'host_acceptance_rate',
              'price', 'extra_people', 'security_deposit', 'cleaning_fee']
num_cols = [i for i in quantitative if i not in date_cols+value_cols]

list_cols = ['host_verifications', 'amenities']

In [521]:
def clean_data(df, drop_var):
    
    df = df.dropna(subset=['price'], axis=0)
    
    for col in date_cols:
        for i in df[col].index:
            if not pd.isnull(df[col][i]):
                delta = date.today() - date.fromisoformat(df[col][i])
                df.loc[i, col] = delta.days
        df.loc[:, col] = df[col].astype(float)

    for col in value_cols:
        for i in df[col].index:
            if not pd.isnull(df[col][i]):
                df.loc[i, col] = re.sub("[$,%]", "", df[col][i])
        df.loc[:, col] = df[col].astype(float)

    for col in list_cols:
        for i in df[col].index:
            if not pd.isnull(df[col][i]):
                if '{' in df[col][i]:
                    df[col][i] = df[col][i].strip('{}').split(',')
                else:
                    df[col][i] = ast.literal_eval(df[col][i])
                
    for col in cat_cols:
        try:
            df = pd.concat([df.drop(col, axis=1), pd.get_dummies(df[col], 
                 prefix=col, prefix_sep='_', dummy_na=False, drop_first=True)], axis=1)
        except:
            df = df.drop(col, axis=1)
            continue

            
    num_vars = df.select_dtypes(include=['float', 'int']).columns
    nan_zero_cols = ['cleaning_fee', 'extra_people']
    
    for col in num_vars:
        if col in nan_zero_cols:
            df.loc[:, col].fillna(0, inplace=True)
        else:
            df.loc[:, col].fillna((df[col].mean()), inplace=True)
    
    y = df['price']
    df = df.drop(drop_var, axis=1)    
    X = df

    return X, y
    
#Use the function to create X and y
drop_list = ['id', 'listing_url', 'scrape_id', 'last_scraped', 'name', 'summary', 
             'experiences_offered', 'host_about','neighborhood_overview', 'space', 'price', 'description', 
             'notes', 'transit', 'thumbnail_url', 'medium_url', 'picture_url', 'xl_picture_url',
             'host_id', 'host_url', 'host_name', 'square_feet', 'host_thumbnail_url', 'host_picture_url',
             'zipcode', 'market', 'smart_location', 'country', 'latitude', 'longitude', 'weekly_price',
             'monthly_price', 'calendar_updated', 'calendar_last_scraped', 'license', 'jurisdiction_names', 'street',
             'neighbourhood', 'neighbourhood_group_cleansed', 'host_neighbourhood', 'host_location'] 

X, y = clean_data(list_df, drop_list)    

/Users/justinharrison/thesis/env/lib/python3.7/site-packages/ipykernel_launcher.py:24: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
/Users/justinharrison/thesis/env/lib/python3.7/site-packages/ipykernel_launcher.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy


In [522]:
X.head()

host_since  host_response_rate  host_acceptance_rate  host_listings_count  \
0      3351.0           96.000000            100.000000                  3.0   
1      2791.0           98.000000            100.000000                  6.0   
2      2315.0           67.000000            100.000000                  2.0   
3      2533.0           94.886798             99.967159                  1.0   
4      3241.0          100.000000             99.967159                  2.0   

   host_total_listings_count  accommodates  bathrooms  bedrooms  beds  \
0                        3.0             4        1.0       1.0   1.0   
1                        6.0             4        1.0       1.0   1.0   
2                        2.0            11        4.5       5.0   7.0   
3                        1.0             3        1.0       0.0   2.0   
4                        2.0             6        2.0       3.0   3.0   

   security_deposit  ...  neighbourhood_cleansed_Windermere  \
0        287.610397  ...                                  0   
1        100.000000  ...                                  0   
2       1000.000000  ...                                  0   
3        287.610397  ...                                  0   
4        700.000000  ...                                  0   

   neighbourhood_cleansed_Yesler Terrace  city_Phinney Ridge Seattle  \
0                                      0                           0   
1                                      0                           0   
2                                      0                           0   
3                                      0                           0   
4                                      0                           0   

   city_Seattle  city_Seattle   city_West Seattle  city_seattle  city_西雅图  \
0             1              0                  0             0         0   
1             1              0                  0             0         0   
2             1              0                  0             0         0   
3             1              0                  0             0         0   
4             1              0                  0             0         0   

   state_wa  is_location_exact_t  
0         0                    1  
1         0                    1  
2         0                    1  
3         0                    1  
4         0                    1  

[5 rows x 157 columns]

In [523]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = .30, random_state=40) 

lm_model = LinearRegression(normalize=True) # Instantiate
lm_model.fit(X_train, y_train) #Fit
        
#Predict and score the model
y_test_preds = lm_model.predict(X_test) 
"The r-squared score for your model was {} on {} values.".format(r2_score(y_test, y_test_preds), len(y_test))

'The r-squared score for your model was -1.1935523632175966e+27 on 1146 values.'

In [524]:
def coef_weights(coefficients, X_train):
    '''
    INPUT:
    coefficients - the coefficients of the linear model 
    X_train - the training data, so the column names can be used
    OUTPUT:
    coefs_df - a dataframe holding the coefficient, estimate, and abs(estimate)
    
    Provides a dataframe that can be used to understand the most influential coefficients
    in a linear model by providing the coefficient estimates along with the name of the 
    variable attached to the coefficient.
    '''
    coefs_df = pd.DataFrame()
    coefs_df['est_int'] = X_train.columns
    coefs_df['coefs'] = lm_model.coef_
    coefs_df['abs_coefs'] = np.abs(lm_model.coef_)
    coefs_df = coefs_df.sort_values('abs_coefs', ascending=False)
    return coefs_df

#Use the function
coef_df = coef_weights(lm_model.coef_, X_train)

#A quick look at the top results
coef_df

est_int         coefs     abs_coefs
150                city_Seattle  6.073103e+16  6.073103e+16
152           city_West Seattle  6.073103e+16  6.073103e+16
151               city_Seattle   6.073103e+16  6.073103e+16
153                city_seattle  3.464852e+16  3.464852e+16
155                    state_wa  2.608251e+16  2.608251e+16
..                          ...           ...           ...
20                 first_review -1.022419e-02  1.022419e-02
0                    host_since -1.120979e-03  1.120979e-03
14               maximum_nights  5.315897e-04  5.315897e-04
149  city_Phinney Ridge Seattle -4.974986e-08  4.974986e-08
154                    city_西雅图 -2.842171e-14  2.842171e-14

[157 rows x 3 columns]

In [506]:
with pd.option_context("display.max_rows", 1000):
    display(coef_df)

est_int         coefs  \
240              neighbourhood_North Beach/Blue Ridge -3.653516e+16   
96          host_neighbourhood_North Beach/Blue Ridge  3.653516e+16   
47            host_neighbourhood_Clairemont Mesa East -2.212986e+16   
178                                     room_type_nan  1.951265e+16   
362                                      city_Seattle -1.942484e+16   
364                                 city_West Seattle -1.942484e+16   
363                                     city_Seattle  -1.942484e+16   
185                              instant_bookable_nan  1.791424e+16   
256                             neighbourhood_Roxhill -1.573797e+16   
113                        host_neighbourhood_Roxhill  1.573797e+16   
223                 neighbourhood_Industrial District  1.548880e+16   
73             host_neighbourhood_Industrial District -1.548880e+16   
51                          host_neighbourhood_Dunlap -1.505400e+16   
207                              neighbourhood_Dunlap  1.505400e+16   
126                  host_neighbourhood_Waianae Coast  1.439224e+16   
50                host_neighbourhood_Downtown Toronto  1.419622e+16   
140                          host_has_profile_pic_nan  1.412455e+16   
188                           cancellation_policy_nan -1.317920e+16   
159                  neighbourhood_group_cleansed_nan  1.297340e+16   
365                                      city_seattle -1.225761e+16   
98              host_neighbourhood_North College Park -1.216026e+16   
190                 require_guest_profile_picture_nan  1.156655e+16   
193                       neighbourhood_Arbor Heights -1.152393e+16   
32                   host_neighbourhood_Arbor Heights  1.152393e+16   
54                      host_neighbourhood_Fauntleroy -1.010495e+16   
117              host_neighbourhood_South Beacon Hill  9.486640e+15   
39                        host_neighbourhood_Brighton -9.014066e+15   
198                            neighbourhood_Brighton  9.014066e+15   
142                        host_identity_verified_nan -8.855205e+15   
125                     host_neighbourhood_View Ridge  7.345690e+15   
368                                          state_wa -7.167232e+15   
111                      host_neighbourhood_Riverview -6.559021e+15   
254                           neighbourhood_Riverview  6.559021e+15   
210                          neighbourhood_Fauntleroy  6.303086e+15   
68                   host_neighbourhood_Highland Park -5.635873e+15   
221                       neighbourhood_Highland Park  5.635873e+15   
351                 neighbourhood_cleansed_View Ridge -5.583167e+15   
138                             host_is_superhost_nan -5.269341e+15   
329              neighbourhood_cleansed_Olympic Hills -4.843590e+15   
150            neighbourhood_group_cleansed_Lake City  4.843590e+15   
250                         neighbourhood_Portage Bay -4.736036e+15   
107                    host_neighbourhood_Portage Bay  4.736036e+15   
165                              property_type_Chalet  4.149991e+15   
203                          neighbourhood_Cedar Park -3.598566e+15   
332                  neighbourhood_cleansed_Pinehurst -3.487551e+15   
316                 neighbourhood_cleansed_Maple Leaf -3.487551e+15   
326         neighbourhood_cleansed_North College Park -3.487551e+15   
152            neighbourhood_group_cleansed_Northgate  3.487551e+15   
300                neighbourhood_cleansed_Haller Lake -3.487551e+15   
234                       neighbourhood_Mathews Beach -3.012812e+15   
70                      host_neighbourhood_Holly Park  2.826722e+15   
222                          neighbourhood_Holly Park -2.826722e+15   
297                 neighbourhood_cleansed_Georgetown -2.728358e+15   
226                         neighbourhood_Laurelhurst -2.696447e+15   
79                     host_neighbourhood_Laurelhurst  2.696447e+15   
67                      host_neighbourhood_High Point  2.644240e+15   

In [208]:
ast.literal_eval(list_df['host_verifications'][0])

['email', 'phone', 'reviews', 'kba']

In [209]:
az_df = list_df['host_verifications']
for i in az_df.index:
    if not pd.isnull(az_df[i]):
        az_df[i] = ast.literal_eval(az_df[i])

/Users/justinharrison/thesis/env/lib/python3.7/site-packages/ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  after removing the cwd from sys.path.


['email', 'phone', 'reviews', 'kba']

In [213]:
new_df = list_df['host_verifications']
mlb = MultiLabelBinarizer()

pd.DataFrame(mlb.fit_transform(az_df),columns=mlb.classes_, index=az_df.index)

TypeError: 'NoneType' object is not iterable

In [214]:
pd.get_dummies(az_df.apply(pd.Series).stack()).sum(level=0)

amex  email  facebook  google  jumio  kba  linkedin  manual_offline  \
0        0      1         0       0      0    1         0               0   
1        0      1         1       0      1    0         1               0   
2        0      1         0       1      1    0         0               0   
3        0      1         1       0      1    0         0               0   
4        0      1         1       0      0    1         0               0   
...    ...    ...       ...     ...    ...  ...       ...             ...   
3813     0      1         0       0      1    0         1               0   
3814     0      1         1       0      1    0         0               0   
3815     0      0         0       0      0    0         0               0   
3816     0      1         1       0      0    1         0               0   
3817     0      1         0       0      0    1         0               0   

      manual_online  phone  photographer  reviews  sent_id  weibo  
0                 0      1             0        1        0      0  
1                 0      1             0        1        0      0  
2                 0      1             0        1        0      0  
3                 0      1             0        1        0      0  
4                 0      1             0        1        0      0  
...             ...    ...           ...      ...      ...    ...  
3813              0      1             0        1        0      0  
3814              0      1             0        1        0      0  
3815              0      1             0        0        0      0  
3816              0      1             0        1        0      0  
3817              0      1             0        1        0      0  

[3815 rows x 14 columns]

In [215]:
amen_ls = list_df[categorical]['amenities'][0].strip('{}').split(',')
for x in amen_ls:
    x = x.strip('""')
amen_ls

['TV',
 '"Cable TV"',
 'Internet',
 '"Wireless Internet"',
 '"Air Conditioning"',
 'Kitchen',
 'Heating',
 '"Family/Kid Friendly"',
 'Washer',
 'Dryer']

In [221]:
list_df['host_verifications'][4].strip('{}')

"['email', 'phone', 'facebook', 'reviews', 'kba']"

In [234]:
others = [i for i in categorical if i not in list_cols]


host_location  host_response_time  \
0     Seattle, Washington, United States  within a few hours   
1     Seattle, Washington, United States      within an hour   
2     Seattle, Washington, United States  within a few hours   
3     Seattle, Washington, United States                 NaN   
4     Seattle, Washington, United States      within an hour   
...                                  ...                 ...   
3813                                  US  within a few hours   
3814  Seattle, Washington, United States      within an hour   
3815                                  US                 NaN   
3816   Tacoma, Washington, United States      within an hour   
3817  Seattle, Washington, United States        within a day   

     host_is_superhost host_has_profile_pic host_identity_verified  \
0                    f                    t                      t   
1                    t                    t                      t   
2                    f                    t                      t   
3                    f                    t                      t   
4                    f                    t                      t   
...                ...                  ...                    ...   
3813                 f                    t                      t   
3814                 f                    t                      t   
3815                 f                    t                      f   
3816                 f                    t                      t   
3817                 f                    t                      t   

     neighbourhood_group_cleansed property_type        room_type  bed_type  \
0                      Queen Anne     Apartment  Entire home/apt  Real Bed   
1                      Queen Anne     Apartment  Entire home/apt  Real Bed   
2                      Queen Anne         House  Entire home/apt  Real Bed   
3                      Queen Anne     Apartment  Entire home/apt  Real Bed   
4                      Queen Anne         House  Entire home/apt  Real Bed   
...                           ...           ...              ...       ...   
3813          Other neighborhoods         House  Entire home/apt  Real Bed   
3814                 Capitol Hill     Apartment  Entire home/apt  Real Bed   
3815               Rainier Valley         House  Entire home/apt  Real Bed   
3816                 Capitol Hill   Condominium  Entire home/apt  Real Bed   
3817                   Queen Anne     Apartment  Entire home/apt  Real Bed   

     instant_bookable cancellation_policy require_guest_profile_picture  \
0                   f            moderate                             f   
1                   f              strict                             t   
2                   f              strict                             f   
3                   f            flexible                             f   
4                   f              strict                             f   
...               ...                 ...                           ...   
3813                f              strict                             f   
3814                f            moderate                             f   
3815                f            moderate                             f   
3816                f            moderate                             f   
3817                f            flexible                             f   

     require_guest_phone_verification  
0                                   f  
1                                   t  
2                                   f  
3                                   f  
4                                   f  
...                               ...  
3813                                f  
3814                                f  
3815                                f  
3816                                f  
3817                                f  

[3818 rows x 13 columns]

In [228]:
list_df[quantitative]

host_since host_response_rate host_acceptance_rate  host_listings_count  \
0     2011-08-11                96%                 100%                  3.0   
1     2013-02-21                98%                 100%                  6.0   
2     2014-06-12                67%                 100%                  2.0   
3     2013-11-06                NaN                  NaN                  1.0   
4     2011-11-29               100%                  NaN                  2.0   
...          ...                ...                  ...                  ...   
3813  2015-04-13                99%                 100%                354.0   
3814  2015-10-14               100%                 100%                  1.0   
3815  2015-12-30                NaN                  NaN                  1.0   
3816  2015-01-03               100%                  NaN                  1.0   
3817  2014-04-25               100%                  NaN                  1.0   

      host_total_listings_count  accommodates  bathrooms  bedrooms  beds  \
0                           3.0             4        1.0       1.0   1.0   
1                           6.0             4        1.0       1.0   1.0   
2                           2.0            11        4.5       5.0   7.0   
3                           1.0             3        1.0       0.0   2.0   
4                           2.0             6        2.0       3.0   3.0   
...                         ...           ...        ...       ...   ...   
3813                      354.0             6        2.0       3.0   3.0   
3814                        1.0             4        1.0       1.0   2.0   
3815                        1.0             2        1.0       1.0   1.0   
3816                        1.0             2        1.0       0.0   1.0   
3817                        1.0             3        1.5       2.0   1.0   

      square_feet  ... first_review last_review review_scores_rating  \
0             NaN  ...   2011-11-01  2016-01-02                 95.0   
1             NaN  ...   2013-08-19  2015-12-29                 96.0   
2             NaN  ...   2014-07-30  2015-09-03                 97.0   
3             NaN  ...          NaN         NaN                  NaN   
4             NaN  ...   2012-07-10  2015-10-24                 92.0   
...           ...  ...          ...         ...                  ...   
3813          NaN  ...   2015-09-27  2015-09-27                 80.0   
3814          NaN  ...   2015-12-18  2015-12-24                100.0   
3815          NaN  ...          NaN         NaN                  NaN   
3816          NaN  ...          NaN         NaN                  NaN   
3817          NaN  ...          NaN         NaN                  NaN   

      review_scores_accuracy review_scores_cleanliness  review_scores_checkin  \
0                       10.0                      10.0                   10.0   
1                       10.0                      10.0                   10.0   
2                       10.0                      10.0                   10.0   
3                        NaN                       NaN                    NaN   
4                        9.0                       9.0                   10.0   
...                      ...                       ...                    ...   
3813                     8.0                      10.0                    4.0   
3814                    10.0                      10.0                   10.0   
3815                     NaN                       NaN                    NaN   
3816                     NaN                       NaN                    NaN   
3817                     NaN                       NaN                    NaN   

      review_scores_communication  review_scores_location  \
0                            10.0                     9.0   
1                            10.0                    10.0   
2                            10.0                    10.0   
3                             NaN     

In [318]:
exp_df = list_df[cat_cols]
list_df[list_cols]['host_verifications'][0]

"['email', 'phone', 'reviews', 'kba']"

In [299]:
exp_df['host_verifications'][0]

"['email', 'phone', 'reviews', 'kba']"

In [341]:
target = list_df[list_cols]
target['host_verifications'][0]

"['email', 'phone', 'reviews', 'kba']"

In [342]:
for i in target['host_verifications'].index:
    if not pd.isnull(target['host_verifications'][i]):
        if '{' in target['host_verifications'][i]:
            print('oops')
            target.loc[i, 'host_verifications'] = target['host_verifications'][i].strip('{}').split(',')
        target.loc[i, 'host_verifications'] = ast.literal_eval(target['host_verifications'][i])

/Users/justinharrison/thesis/env/lib/python3.7/site-packages/pandas/core/indexing.py:670: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_with_indexer(indexer, value)
/Users/justinharrison/thesis/env/lib/python3.7/site-packages/ipykernel_launcher.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


In [350]:
for x in target['host_verifications']:
    print(type(x) )

<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'li

<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'li

<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'list'>
<class 'li

In [345]:
target['host_verifications'][i]

['email', 'phone', 'reviews', 'kba']